Document required fd/socket flags for I/O safety.

sunfishcode opened the pull request:
On Unix, document that the `O_CLOEXEC` flag is required on all file descriptors, and on Windows, document that the `WSA_FLAG_OVERLAPPED` and `WSA_FLAG_NO_HANDLE_INHERIT` flags are required on all sockets, as a as a requirement for I/O safety. `O_CLOEXEC` and
`WSA_FLAG_NO_HANDLE_INHERIT` prevent resources from being implicitly leaked into child processes, and `WSA_FLAG_OVERLAPPED` prevents undefined behavior in usage patterns which safe Rust ought to be able to depend on.

The Rust standard library already sets these flags on all file descriptors and sockets it creates, so there's no need to change any code in std for this. And, this behavior is already a common expectation of crates in the ecosystem, specifically because it avoids the problems mentioned here.

Rustix is one crate that currently creates file descriptors without guaranteeing that `O_CLOEXEC` is set, because it was originally conveived of as a low-level API where `O_CLOEXEC` isn't its responsibility. However, as a rustix author, I now think it makes sense include `O_CLOEXEC` in as a part of the task of providing a safe API, and am submitting this PR to seek broader consensus on this.

[rustix will follow Rust's decision here]: https://github.com/bytecodealliance/rustix/issues/495

sunfishcode committed:
Document required fd/socket flags for I/O safety.

On Unix, document that the `O_CLOEXEC` flag is required on all file
descriptors, and on Windows, document that the `WSA_FLAG_OVERLAPPED` and
`WSA_FLAG_NO_HANDLE_INHERIT` flags are required on all sockets, as a
as a requirement for I/O safety. `O_CLOEXEC` and
`WSA_FLAG_NO_HANDLE_INHERIT` prevent resources from being implicitly
leaked into child processes, and `WSA_FLAG_OVERLAPPED` prevents
undefined behavior in usage patterns which safe Rust ought to be able
to depend on.

The Rust standard library already sets these flags on all file descriptors
and sockets it creates, so there's no need to change any code in std for
this. And, this behavior is already a common expectation of crates in the
ecosystem, specifically because it avoids the problems mentioned here.

Rustix is one crate that currently creates file descriptors without
guaranteeing that `O_CLOEXEC` is set, because it was originally conveived
of as a low-level API where `O_CLOEXEC` isn't it's responsibility. However,
as a rustix author, I now think it makes sense include `O_CLOEXEC` in as a
part of the task of providing a safe API, and am submitting this PR to
seek broader consensus on this.

[rustix will follow Rust's decision here]: https://github.com/bytecodealliance/rustix/issues/495

notgull commented:
Looks good to me; might be a breaking change though. Does this mean that code like [this](https://docs.rs/socket2/latest/socket2/struct.Socket.html#method.new_raw)  will become unsafe?

sunfishcode commented:
> Looks good to me; might be a breaking change though. Does this mean that code like [this](https://docs.rs/socket2/latest/socket2/struct.Socket.html#method.new_raw) will become unsafe?

Yes. The `new` function sets `O_CLOEXEC`  but the `new_raw` function does not, so the proposal here would oblige it to be marked `unsafe`.

inquisitivecrystal commented on library/std/src/os/unix/io/mod.rs:
```diff
@@ -54,6 +54,16 @@
+//! ## Required file-descriptor flags
+//!
+//! On Unix platforms, file descriptors that don't have the `O_CLOEXEC` flag
+//! set may be implicitly leaked into spawned child processes. This can be
```
```suggestion
//! set may be implicitly leaked into spawned child processes. This can
```

inquisitivecrystal commented on library/std/src/os/windows/io/mod.rs:
```diff
@@ -45,6 +45,19 @@
+//!
+//! On Windows, sockets that don't have the `WSA_FLAG_OVERLAPPED` flag can't be
+//! passed to `WSASend` calls when a non-NULL `lpOverlapped` argument is
+//! provided. And sockets that don't have the `WSA_FLAG_NO_HANDLE_INHERIT`
```
```suggestion
//! provided. Additionally, sockets that don't have the `WSA_FLAG_NO_HANDLE_INHERIT`
```

inquisitivecrystal commented on library/std/src/os/windows/io/mod.rs:
```diff
@@ -45,6 +45,19 @@
+//! On Windows, sockets that don't have the `WSA_FLAG_OVERLAPPED` flag can't be
+//! passed to `WSASend` calls when a non-NULL `lpOverlapped` argument is
+//! provided. And sockets that don't have the `WSA_FLAG_NO_HANDLE_INHERIT`
+//! flag set may be implicitly leaked into spawned child processes. This can be
```
```suggestion
//! flag set may be implicitly leaked into spawned child processes. This can
```

inquisitivecrystal commented on library/std/src/os/unix/io/mod.rs:
```diff
@@ -54,6 +54,16 @@
+//! is not intended to be exposed to child processes, and a spawn in another
+//! part of the code unknowingly exposes it. Consequently, Rust code that
+//! produces new `OwnedFd` or `BorrowedFd` values must either set the
+//! `O_CLOEXEC` flag, or be marked `unsafe`.
```
I feel like there's some way of putting this more precisely (something about invariants or the like?) but I don't know what it is. (The same applies to the Windows documentation, if any change is made.)

inquisitivecrystal commented on library/std/src/os/unix/io/mod.rs:
```diff
@@ -54,6 +54,16 @@
+//!
+//! On Unix platforms, file descriptors that don't have the `O_CLOEXEC` flag
+//! set may be implicitly leaked into spawned child processes. This can be
+//! violate I/O safety, when the a file descriptor held in one part of the code
```
```suggestion
//! violate I/O safety, when a file descriptor held in one part of the code
```

inquisitivecrystal commented on library/std/src/os/windows/io/mod.rs:
```diff
@@ -45,6 +45,19 @@
+//! passed to `WSASend` calls when a non-NULL `lpOverlapped` argument is
+//! provided. And sockets that don't have the `WSA_FLAG_NO_HANDLE_INHERIT`
+//! flag set may be implicitly leaked into spawned child processes. This can be
+//! violate I/O safety, when the a socket held in one part of the code is not
```
```suggestion
//! violate I/O safety, when a socket held in one part of the code is not
```

Thomasdezeeuw commented:
I'm coming from https://github.com/rust-lang/socket2/issues/384. I wanted to note that although I'm not against this change, it is a breaking one, one with real breakage as seen in socket2.

sunfishcode committed:
Some OS's don't have spawning, and some can't do atomic O_CLOEXEC.

Acccording to library/std/src/sys/unix/fd.rs, ESP-IDF and Horizon OS
don't have the ability to spawn processes and thus don't need to set
`O_CLOEXEC`.

Also, add a paragraph mentioning that some OS's can't set `O_CLOEXEC`
atomically, and so it should be done as-soon-as-possible.

sunfishcode commented on library/std/src/os/unix/io/mod.rs:
```diff
@@ -54,6 +54,23 @@
+//! child processes, and a spawn in another part of the code unknowingly
+//! exposes it. Consequently, on these platforms, Rust code that produces new
+//! `OwnedFd` or `BorrowedFd` values must either set the `O_CLOEXEC` flag, or
+//! be marked `unsafe`.
```
If you get an `OwnedFd` from an unknown source (eg. an argument to a library API), I don't think we can give you any kind of guarantee that `O_CLOEXEC` is set on it. We don't give you a guarantee that there are no `dup`'d copies, for example, and that's conceptaully a very similar situation. `fork` copies file descriptors in the same manner as `dup`, sharing file descriptions.

My intention with "marked `unsafe`" is to say that there must be an `unsafe` just at the point where the first `OwnedFd` or `BorrowedFd` is created. Once you have obtained an `OwnedFd` or `BorrowedFd` with a non-`O_CLOEXEC` fd in it though, it's ok to implement `AsFd` and hand out new `BorrowedFd`'s without `unsafe`. So `new_raw` would need to be `unsafe`, but it looks like the rest of the `Socket` API could remain as-is. I'll work on better wording for this.

Thomasdezeeuw replied:
> If you get an `OwnedFd` from an unknown source (eg. an argument to a library API), I don't think we can give you any kind of guarantee that `O_CLOEXEC` is set on it. We don't give you a guarantee that there are no `dup`'d copies, for example, and that's conceptaully a very similar situation. `fork` copies file descriptors in the same manner as `dup`, sharing file descriptions.

I agree, but then I'm even more confused as what I should read into this paragraph of documentation.

> My intention with "marked `unsafe`" is to say that there must be an `unsafe` just at the point where the first `OwnedFd` or `BorrowedFd` is created. Once you have obtained an `OwnedFd` or `BorrowedFd` with a non-`O_CLOEXEC` fd in it though, it's ok to implement `AsFd` and hand out new `BorrowedFd`'s without `unsafe`. So `new_raw` would need to be `unsafe`, but it looks like the rest of the `Socket` API could remain as-is. I'll work on better wording for this.

But why should the creation function be `unsafe`? I'm failing to see what benefit it has.

sunfishcode commented:
As discussed above, I now think this PR is the wrong approach.

sunfishcode commented:
It'd be too restrictive to say that `OwnedFd` always requires its fd to have `O_CLOEXEC`. There are important use cases that unset `O_CLOEXEC`. That led me to thinking that perhaps `FromRawFd::from_raw_fd` and `BorrowedFd::borrow_raw` should require `O_CLOEXEC`, but that also would preclude meaningful use cases.

I think the way to think about `O_CLOEXEC` is that if a fd ever exists without `O_CLOEXEC`, then whatever code created it that way, or caused it to be that way, has conceptually added the fd to a global set of escaped fds, which is read when a new process is spawned. It's still desirable for libraries to set `O_CLOEXEC` for new fds, but it doesn't break encapsulation if they don't.
